### PR TITLE
Feat/settlement

### LINKED ISSUE #10: Generate a Settlement Plan
As a group member,

I want the app to calculate and display a simplified settlement plan,

So that everyone can pay off their debts with the fewest possible transactions.

### Acceptance Criteria

- [ ] **AC1 — Settlement plan minimises transactions**
  Given a group has multiple members with non-zero net balances
  When a member opens the "Settlement plan" view
  Then the app displays a list of payments (from, to, amount) that clears all balances
  And the number of payments is the minimum possible, not one payment per original expense

- [ ] **AC2 — All balances reach zero after the plan**
  Given the settlement plan shown for a group
  When every listed payment is applied
  Then every member's net balance becomes 0

- [ ] **AC3 — Amounts are correctly directed**
  Given the settlement plan is displayed
  When a payment line is shown
  Then it reads as a specific debtor paying a specific creditor a specific amount, never a negative or ambiguous figure

- [ ] **AC4 — Already-settled group**
  Given all members in a group currently have a net balance of 0
  When a member opens the settlement plan view
  Then no payments are listed
  And a message states the group is already settled

- [ ] **AC5 — Marking a payment as settled**
  Given a settlement plan lists a payment from user A to user B
  When user A or user B marks that payment as paid
  Then that debt is removed from both users' outstanding balances
  And the settlement plan recalculates for the remaining unsettled balances

- [ ] **AC6 — Recalculates after new expenses**
  Given a settlement plan has been generated
  When a new expense is added to the group before the plan is fully settled
  Then reopening the settlement plan view reflects the updated balances

- [ ] **AC7 — Single-member or single-debt edge case**
  Given only two members in a group have non-zero balances and they exactly offset each other
  When the settlement plan is generated
  Then exactly one payment is listed, from the debtor to the creditor

- [ ] **AC8 — Only members can view or settle**
  Given a user is not a member of the group
  When they attempt to view or mark payments in its settlement plan via the API or UI
  Then the request is rejected with an authorisation error

### PR DESCRIPTION
**Pull request by AdRo-Coder**
## Summary
Added calculation logic for the settlement plan.

## Associated issue
Closes #10 

## Changes
- Calculation of the settlement plan
- Updating balance view

## Testing
Tests run locally in IDE

## Dependencies
-

## Screenshots
image

image

## Reviewer notes
logic of settlement plan.

## Generative AI
AI was used for helping with coding and writing test. AI used: Copilot

## Checklist

- [x] Associated with an open, team-approved issue
- [x] Branched from `main` and rebased against it
- [x] Test suite passes and the application runs
- [x] Tests are added alongside any new or modified code
- [x] No unrelated changes are included
- [x] Documentation updated where the change requires it
- [x] Any use of generative AI is acknowledged above

### CODE REVIEW

**Comment by yd35** [comment]
Had a look before review. A few things need fixing before this can merge:

- PR #54 and this PR both add a V5 Flyway migration. Since Issue #10 depends on Issue #8, PR #54 should merge first. This branch then needs to be rebased and the settlement migration renamed to V6.
- The settlement calculation currently rebuilds balances by dividing the total spend equally across every group member. PR #54 allows selected participants, so this needs to use the stored member balances or expense-share data. Otherwise subset splits will produce the wrong settlement plan.
- computeBalancesIncludingSettlements currently applies open settlements and skips paid settlements. This makes a balance look cleared before payment, then brings it back after Mark as paid. It should apply completed payments instead.

Once those are fixed, please rerun the full backend and frontend tests and the app.

File: backend/src/main/java/nz/ac/auckland/se310/fairshare/service/ExpenseGroupService.java [hunk 1]
```diff
@@ -102,6 +121,202 @@ public void removeMember(Long groupId, Long memberUserId, Long currentUserId) {
...
+
+            if (debtor.amount.compareTo(new BigDecimal("0.005")) < 0) i++;
+            if (creditor.amount.compareTo(new BigDecimal("0.005")) < 0) j++;
+        }
+
+        return transactions;
+    }
+
+    private Map<Long, BigDecimal> computeBalancesIncludingSettlements(ExpenseGroup group) {
+        Map<Long, BigDecimal> balances = new HashMap<>();
```
Comment (yd35) [inline backend/src/main/java/nz/ac/auckland/se310/fairshare/service/ExpenseGroupService.java]: This starts from an empty map and only adds open settlements. Right after an expense is added, there is no open settlement, so `/members` and `/balances` return zero for everyone. `SettlementView` then reports that everyone is settled and never generates the first plan. My MySQL regression test returned `{alice=0, bob=0}` instead of `{alice=21.25, bob=-21.25}`. Start from the stored member balances or exact expense shares and restore the deleted balance tests.

File: backend/src/main/java/nz/ac/auckland/se310/fairshare/service/ExpenseGroupService.java [hunk 2]
```diff
@@ -102,6 +121,202 @@ public void removeMember(Long groupId, Long memberUserId, Long currentUserId) {
...
+            for (Long participantId : participantIds) {
+                balances.merge(participantId, sharePerParticipant, BigDecimal::add);
+            }
+        }
+
+        for (Settlement settlement : settlementRepository.findByGroupId(groupId)) {
+            if (settlement.getSettlementDate() != null) {
+                Long fromId = settlement.getFromUser().getId();
+                Long toId = settlement.getToUser().getId();
+                balances.merge(fromId, settlement.getAmount(), BigDecimal::add);
```
Comment (yd35) [inline backend/src/main/java/nz/ac/auckland/se310/fairshare/service/ExpenseGroupService.java:231]: The completed payment is applied in the wrong direction. With positive meaning owes, the payment must subtract from `fromUser` and add to `toUser`. A paid NZD 50 debt currently recalculates as a new NZD 100 debt instead of clearing it.

File: backend/src/main/java/nz/ac/auckland/se310/fairshare/service/ExpenseGroupService.java [hunk 3]
```diff
@@ -102,6 +121,202 @@ public void removeMember(Long groupId, Long memberUserId, Long currentUserId) {
...
+                balances.merge(toId, settlement.getAmount().negate(), BigDecimal::add);
+            }
+        }
+
+        return balances;
+    }
+
+    private void persistSettlementPlan(ExpenseGroup group, Long groupId, List<SettlementLine> settlementPlan) {
```
Comment (yd35) [inline backend/src/main/java/nz/ac/auckland/se310/fairshare/service/ExpenseGroupService.java:244]: This only updates pairs present in the new plan. If the new plan is empty or a pair disappears, the old open settlement is never removed, so stale debt remains. Reconcile or replace the complete set of open settlements on every recalculation.

File: backend/src/main/java/nz/ac/auckland/se310/fairshare/service/ExpenseGroupService.java [hunk 4]
```diff
@@ -102,6 +121,202 @@ public void removeMember(Long groupId, Long memberUserId, Long currentUserId) {
...
+        for (Map.Entry<Long, BigDecimal> entry : totalExpenses.entrySet()) {
+            BigDecimal amount = entry.getValue() == null ? BigDecimal.ZERO : entry.getValue();
+            if (amount.compareTo(BigDecimal.ZERO) > 0) {
+                debtors.add(new PersonBalance(entry.getKey(), amount));
+            } else if (amount.compareTo(BigDecimal.ZERO) < 0) {
+                creditors.add(new PersonBalance(entry.getKey(), amount.abs()));
+            }
+        }
+
+        debtors.sort(Comparator.comparingLong(balance -> balance.name));
```
Comment (yd35) [inline backend/src/main/java/nz/ac/auckland/se310/fairshare/service/ExpenseGroupService.java]: Sorting by user ID and pairing in order does not guarantee the minimum number of payments required by AC1. For balances `+5, +10, -10, -5`, this returns three payments, but matching the equal amounts only needs two.

File: backend/src/main/java/nz/ac/auckland/se310/fairshare/service/ExpenseGroupService.java [hunk 5]
```diff
@@ -102,6 +121,202 @@ public void removeMember(Long groupId, Long memberUserId, Long currentUserId) {
...
+
+        Map<Long, BigDecimal> effectiveBalances = computeEffectiveBalances(groupId, group, expenseService.getExpensesForGroup(groupId, currentUserId));
+        List<SettlementLine> settlementPlan = calculateSettlements(effectiveBalances);
+        persistSettlementPlan(group, groupId, settlementPlan);
+        return settlementPlan;
+    }
+
+    @Transactional
+    public void markSettlementPaid(Long groupId, Long fromUserId, Long toUserId, Long currentUserId) {
+        requireMemberGroup(groupId, currentUserId);
```
Comment (yd35) [inline backend/src/main/java/nz/ac/auckland/se310/fairshare/service/ExpenseGroupService.java]: AC5 only allows the payer or recipient to mark this payment as paid. This currently allows any group member to do it. Check that `currentUserId` matches `fromUserId` or `toUserId`, and hide the button from other members in the UI.

File: frontend/src/pages/GroupPage.jsx [hunk 6]
```diff
@@ -12,10 +12,7 @@ function money(currency, value) {
 function balanceLine(member, currency) {
     const balance = Number(member.netBalance);
     if (balance > 0) {
-        return `${member.username} is owed ${money(currency, balance)}`;
-    }
-    if (balance < 0) {
-        return `${member.username} owes ${money(currency, -balance)}`;
+        return `${member.username} owes ${money(currency, balance)}`;
```
Comment (yd35) [inline frontend/src/pages/GroupPage.jsx:15]: With the new sign convention, negative balances mean the member is owed money, but they now fall through to `is settled up`. Restore an `is owed` branch and keep the existing UI regression test for both directions.

**Review by yd35 — CHANGES_REQUESTED** [review]
Thanks, the branch is now rebased, the migration is V6, and the normal test suites pass. I ran 52 backend tests, 55 frontend tests, the frontend build, and focused regression tests. The focused tests exposed the blockers listed inline. Please fix these cases, restore the deleted expense balance and split tests, and add regression coverage for Issue #10. I can re-review quickly once pushed.

File: frontend/src/pages/GroupPage.jsx (hunk 6 above)
Comment (AdRo-Coder) [reply frontend/src/pages/GroupPage.jsx:15]: image
image

Negative means you owe money and it's marked red.

File: backend/src/main/java/nz/ac/auckland/se310/fairshare/service/ExpenseGroupService.java (hunk 5 above)
Comment (AdRo-Coder) [reply backend/src/main/java/nz/ac/auckland/se310/fairshare/service/ExpenseGroupService.java]: Resolved in new commit.

File: backend/src/main/java/nz/ac/auckland/se310/fairshare/service/ExpenseGroupService.java (hunk 4 above)
Comment (AdRo-Coder) [reply backend/src/main/java/nz/ac/auckland/se310/fairshare/service/ExpenseGroupService.java]: Resolved in new commit.

File: backend/src/main/java/nz/ac/auckland/se310/fairshare/service/ExpenseGroupService.java (hunk 3 above)
Comment (AdRo-Coder) [reply backend/src/main/java/nz/ac/auckland/se310/fairshare/service/ExpenseGroupService.java:244]: Resolved in new commit.

File: backend/src/main/java/nz/ac/auckland/se310/fairshare/service/ExpenseGroupService.java (hunk 2 above)
Comment (AdRo-Coder) [reply backend/src/main/java/nz/ac/auckland/se310/fairshare/service/ExpenseGroupService.java:231]: Resolved in new commit.

File: backend/src/main/java/nz/ac/auckland/se310/fairshare/service/ExpenseGroupService.java (hunk 1 above)
Comment (AdRo-Coder) [reply backend/src/main/java/nz/ac/auckland/se310/fairshare/service/ExpenseGroupService.java]: Resolved in new commit.

File: backend/src/main/java/nz/ac/auckland/se310/fairshare/service/ExpenseGroupService.java [hunk 7]
```diff
@@ -102,6 +121,248 @@ public void removeMember(Long groupId, Long memberUserId, Long currentUserId) {
         groupRepository.save(group);
     }
 
+    @Transactional(readOnly = true)
+    public List<MemberBalance> getBalances(Long groupId, Long currentUserId) {
+        ExpenseGroup group = requireMemberGroup(groupId, currentUserId);
+        Map<Long, BigDecimal> memberBalances = computeEffectiveBalances(groupId, group, null);
```
Comment (yd35) [inline backend/src/main/java/nz/ac/auckland/se310/fairshare/service/ExpenseGroupService.java]: This still passes `null`, so `computeEffectiveBalances` ignores the saved expense balances and returns 0 for every member before any payment. `SettlementView` then says everyone is settled until the user clicks Generate. Issue #10 AC1 says the plan should be shown when the view opens. Please use the current expenses here or start from the stored balances.

File: backend/src/main/java/nz/ac/auckland/se310/fairshare/service/ExpenseGroupService.java [hunk 8]
```diff
@@ -102,6 +121,248 @@ public void removeMember(Long groupId, Long memberUserId, Long currentUserId) {
...
+        Long recipientId = settlement.getToUser().getId();
+
+        UserInGroup payer = group.getMember(payerId);
+        UserInGroup recipient = group.getMember(recipientId);
+
+        BigDecimal amount = settlement.getAmount() == null ? BigDecimal.ZERO : settlement.getAmount();
+
+        if (payer != null) {
+            BigDecimal old = payer.getNetBalance() == null ? BigDecimal.ZERO : payer.getNetBalance();
+            payer.adjustNetBalance(old.subtract(amount));
```
Comment (yd35) [inline backend/src/main/java/nz/ac/auckland/se310/fairshare/service/ExpenseGroupService.java]: `adjustNetBalance` takes a change in balance. For a payer on +50, `old.subtract(amount)` is 0, so adding it leaves the payer on +50. The recipient has the same problem. My regression test left both stored balances unchanged after marking the payment paid. Apply `-amount` to the payer and `+amount` to the recipient so AC5 and the member removal check work.

File: backend/src/main/java/nz/ac/auckland/se310/fairshare/service/ExpenseGroupService.java [hunk 9]
```diff
@@ -102,6 +121,248 @@ public void removeMember(Long groupId, Long memberUserId, Long currentUserId) {
...
+            }
+        }
+
+        balances.replaceAll((key, value) -> value == null ? BigDecimal.ZERO.setScale(2) : value.setScale(2, RoundingMode.HALF_UP));
+
+        return balances;
+    }
+
+    private void persistSettlementPlan(ExpenseGroup group, Long groupId, List<SettlementLine> settlementPlan) {
+        for (SettlementLine line : settlementPlan) {
```
Comment (yd35) [inline backend/src/main/java/nz/ac/auckland/se310/fairshare/service/ExpenseGroupService.java:247]: This only checks pairs that exist in the new plan. An old open settlement that is no longer needed is never deleted. My regression test recalculated an empty plan and the old open settlement stayed in the database. Please replace or fully reconcile the open plan so AC6 does not leave stale payments.

File: backend/src/main/java/nz/ac/auckland/se310/fairshare/service/ExpenseGroupService.java [hunk 10]
```diff
@@ -102,6 +121,248 @@ public void removeMember(Long groupId, Long memberUserId, Long currentUserId) {
...
+            if (amount.compareTo(BigDecimal.ZERO) > 0) {
+                debtors.add(new PersonBalance(entry.getKey(), amount));
+            } else if (amount.compareTo(BigDecimal.ZERO) < 0) {
+                creditors.add(new PersonBalance(entry.getKey(), amount.abs()));
+            }
+        }
+
+        // Sort by amount descending to pair largest debtors with largest creditors first —
+        // this tends to minimise the number of transactions compared to arbitrary ID ordering.
+        debtors.sort((a, b) -> b.amount.compareTo(a.amount));
```
Comment (yd35) [inline backend/src/main/java/nz/ac/auckland/se310/fairshare/service/ExpenseGroupService.java]: Largest-first is still a greedy approach and does not guarantee the minimum required by AC1. For balances `+2, +3, -1, -2, -2`, this code creates 4 payments, but 3 payments are enough: `2 -> 2`, then `3 -> 1` and `3 -> 2`. Please add this as a regression test and calculate the actual minimum.

File: frontend/src/pages/GroupPage.jsx [hunk 11]
```diff
@@ -12,10 +12,7 @@ function money(currency, value) {
...
     }
     return `${member.username} is settled up`;
```
Comment (yd35) [inline frontend/src/pages/GroupPage.jsx:19]: Negative balances now mean the member is owed money, but every negative value reaches this line and is shown as settled. For example, `-21.25` displays `alice is settled up`. Please add a negative branch that shows the member is owed the absolute amount.

**Review by yd35 — CHANGES_REQUESTED** [review]
Re-reviewed the latest commit. The authorization and expense-share fixes look good. I ran 57 backend tests, 55 frontend tests and the frontend build successfully. Four focused regression checks still fail, and the negative balance display is still wrong. Details are inline. Please also restore `ac4_unevenSplitStillSumsToTheAmount` and add regression tests for these cases before the next review.

File: frontend/src/pages/GroupPage.jsx (hunk 11 above)
Comment (AdRo-Coder) [reply frontend/src/pages/GroupPage.jsx:19]: resolved in newest commit.

File: backend/src/main/java/nz/ac/auckland/se310/fairshare/service/ExpenseGroupService.java (hunk 10 above)
Comment (AdRo-Coder) [reply backend/src/main/java/nz/ac/auckland/se310/fairshare/service/ExpenseGroupService.java]: resolved in newest commit.

File: backend/src/main/java/nz/ac/auckland/se310/fairshare/service/ExpenseGroupService.java (hunk 9 above)
Comment (AdRo-Coder) [reply backend/src/main/java/nz/ac/auckland/se310/fairshare/service/ExpenseGroupService.java:247]: resolved in newest commit.

File: backend/src/main/java/nz/ac/auckland/se310/fairshare/service/ExpenseGroupService.java (hunk 8 above)
Comment (AdRo-Coder) [reply backend/src/main/java/nz/ac/auckland/se310/fairshare/service/ExpenseGroupService.java]: resolved in newest commit.

File: backend/src/main/java/nz/ac/auckland/se310/fairshare/service/ExpenseGroupService.java (hunk 7 above)
Comment (AdRo-Coder) [reply backend/src/main/java/nz/ac/auckland/se310/fairshare/service/ExpenseGroupService.java]: resolved in newest commit.

File: backend/src/main/java/nz/ac/auckland/se310/fairshare/service/ExpenseGroupService.java [hunk 12]
```diff
@@ -102,6 +121,303 @@ public void removeMember(Long groupId, Long memberUserId, Long currentUserId) {
...
+        Long recipientId = settlement.getToUser().getId();
+
+        UserInGroup payer = group.getMember(payerId);
+        UserInGroup recipient = group.getMember(recipientId);
+
+        BigDecimal amount = settlement.getAmount() == null ? BigDecimal.ZERO : settlement.getAmount();
+
+        if (payer != null) {
+            BigDecimal old = payer.getNetBalance() == null ? BigDecimal.ZERO : payer.getNetBalance();
+            payer.adjustNetBalance(old.add(amount));
```
Comment (yd35) [inline backend/src/main/java/nz/ac/auckland/se310/fairshare/service/ExpenseGroupService.java]: `adjustNetBalance` adds the value passed to it. This passes `old + amount`, so a payer on +50 receives a +100 change and ends on +150. The recipient moves from -50 to -150 as well. My focused test confirmed those results. Pass `amount.negate()` for the payer and `amount` for the recipient so AC5 clears both stored balances.

File: backend/src/main/java/nz/ac/auckland/se310/fairshare/service/ExpenseGroupService.java [hunk 13]
```diff
@@ -102,6 +121,303 @@ public void removeMember(Long groupId, Long memberUserId, Long currentUserId) {
...
+            }
+        }
+
+        balances.replaceAll((key, value) -> value == null ? BigDecimal.ZERO.setScale(2) : value.setScale(2, RoundingMode.HALF_UP));
+
+        return balances;
+    }
+
+    private void persistSettlementPlan(ExpenseGroup group, Long groupId, List<SettlementLine> settlementPlan) {
+        for (SettlementLine line : settlementPlan) {
```
Comment (yd35) [inline backend/src/main/java/nz/ac/auckland/se310/fairshare/service/ExpenseGroupService.java:247]: This previous blocker is unchanged. The loop only checks pairs in the new plan, so an old open settlement remains when the new plan is empty or that pair disappears. My focused test still kept the stale row. Please delete or fully reconcile all open settlements before saving the recalculated plan for AC6.

File: frontend/src/pages/GroupPage.jsx [hunk 14]
```diff
@@ -1,27 +1,26 @@
...
+import {useEffect, useState} from 'react';
+import {Link, useParams} from 'react-router-dom';
+import {getGroup, getGroupMembers} from '../api/groups';
+import {getExpenses} from '../api/expenses';
 import SettlementView from './SettlementView';
 import './GroupPage.css';
 
 function money(currency, value) {
-    return `${currency} ${Number(value).toFixed(2)}`;
+    return `${currency} ${Math.abs(Number(value).toFixed(2))}`;
```
Comment (yd35) [inline frontend/src/pages/GroupPage.jsx]: `toFixed(2)` returns a string, then `Math.abs` converts it back to a number and removes trailing zeros. Amounts now display as `NZD 20` instead of `NZD 20.00`, and the test was changed to accept the regression. Use `Math.abs(Number(value)).toFixed(2)` and keep the original currency-format expectations.

**Review by yd35 — CHANGES_REQUESTED** [review]
Re-reviewed the latest commit. The minimum-transaction case and initial balance load are fixed. I ran 62 backend tests, 55 frontend tests and the frontend build successfully. Three blockers remain and are listed inline. The focused AC5 and AC6 regression tests still fail. Please also restore `ac4_unevenSplitStillSumsToTheAmount` and add regression tests for marking a payment paid and removing stale open settlements.

File: frontend/src/pages/GroupPage.jsx (hunk 14 above)
Comment (AdRo-Coder) [reply frontend/src/pages/GroupPage.jsx]: fixed

**Review by yd35 — CHANGES_REQUESTED** [review]
Currency fix looks good. 55 frontend tests and the build pass. The backend is unchanged, so the open AC5 and AC6 blockers still apply. I cannot approve until those are fixed.

**Review by Jago-Sutherland — APPROVED** [review]
Builds, passes all tests. Did some simple sanity checks and UI appearance checks, all looks good.

File: backend/src/main/java/nz/ac/auckland/se310/fairshare/service/ExpenseGroupService.java [hunk 15]
```diff
@@ -102,6 +121,326 @@ public void removeMember(Long groupId, Long memberUserId, Long currentUserId) {
...
+        Long recipientId = settlement.getToUser().getId();
+
+        UserInGroup payer = group.getMember(payerId);
+        UserInGroup recipient = group.getMember(recipientId);
+
+        BigDecimal amount = settlement.getAmount() == null ? BigDecimal.ZERO : settlement.getAmount();
+
+        if (payer != null) {
+            BigDecimal old = payer.getNetBalance() == null ? BigDecimal.ZERO : payer.getNetBalance();
+            payer.adjustNetBalance(old.add(amount.negate()));
```
Comment (yd35) [inline backend/src/main/java/nz/ac/auckland/se310/fairshare/service/ExpenseGroupService.java]: `adjustNetBalance` expects only the change. This still includes the old balance. In my test Alice went from `-25.00` to `-75.00` and Bob stayed on `25.00` after paying. Use `amount.negate()` for the payer and `amount` for the recipient, then check both end on zero.

File: backend/src/main/java/nz/ac/auckland/se310/fairshare/service/ExpenseGroupService.java [hunk 16]
```diff
@@ -102,6 +121,326 @@ public void removeMember(Long groupId, Long memberUserId, Long currentUserId) {
...
+                continue;
+            }
+
+            Settlement oppositeOpen = findOpenSettlement(groupId, toId, fromId);
+            if (oppositeOpen != null) {
+                BigDecimal oppositeAmount = oppositeOpen.getAmount();
+                int comparison = oppositeAmount.compareTo(amount);
+
+                if (comparison > 0) {
+                    oppositeOpen.setAmount(oppositeAmount.subtract(amount));
```
Comment (yd35) [inline backend/src/main/java/nz/ac/auckland/se310/fairshare/service/ExpenseGroupService.java]: When the new plan reverses direction, this updates the old opposite settlement, then `deleteUnusedOpenSettlements` deletes it because only the new direction is active. My test expected Alice -> Bob `20.00` but no new settlement was saved. Replace the old open settlement with the full new plan and add a test for this case.

**Review by yd35 — CHANGES_REQUESTED** [review]
Re-reviewed the latest commit. The stale settlement cleanup and restored split test look good. The normal backend suite passes with 66 tests. Two focused settlement cases still fail. Details are inline.

File: backend/src/test/java/nz/ac/auckland/se310/fairshare/service/ExpenseGroupServiceSettlementPersistenceTest.java [hunk 17]
```diff
@@ -0,0 +1,250 @@
...
+    @Test
+    void markSettlementPaidSetsSettlementDateAndPersistsPayment() {
+        Settlement open = new Settlement(group, bobUser, aliceUser, new BigDecimal("25.00"));
+        open.setSettlementDate(null);
+        when(settlementRepository.findByGroupIdAndFromUserIdAndToUserIdOrderByIdDesc(GROUP_ID, BOB, ALICE))
+                .thenReturn(List.of(open));
+
+        service.markSettlementPaid(GROUP_ID, BOB, ALICE, BOB);
+
+        assertThat(open.getSettlementDate()).isNotNull();
```
Comment (yd35) [inline backend/src/test/java/nz/ac/auckland/se310/fairshare/service/ExpenseGroupServiceSettlementPersistenceTest.java:198]: The code fix works. This test still only checks the date and save call. Set Alice to `-25.00` and Bob to `25.00` before the call then assert both balances are zero afterwards. This keeps the bug covered.

**Review by yd35 — CHANGES_REQUESTED** [review]
Both fixes look good. I ran all 67 backend tests and the focused payment test. Everything passes. Just add the missing balance assertions to the existing test before merge.

**Review by yd35 — APPROVED** [review]
the requested balance assertions are now included and pass. the settlement tests and all checks pass. lgtm
